[% WRAPPER html_wrapper.html %]

[% genes_on_non_existing_contigs = problem_data.genes_on_non_existing_contigs %]

[% USE dumper %]

<!--
<pre>
[%# dumper.dump(genes_on_non_existing_contigs) %]
</pre>
-->

[% IF genes_on_non_existing_contigs %]

<h1>Missing contigs</h1>

<p>
  There were [% genes_on_non_existing_contigs.size %] genes annotated on contigs that didn't exist in the assembly file:
</p>

<table>
<tr>
  <th>Contig name</th>
  <th>Description</th>
  <th>Start</th>
  <th>End</th>
  <th>Strand</th>
</tr>
[%- FOREACH current_gene = genes_on_non_existing_contigs -%]
<tr>
    <td>[% current_gene.contig_name %]</td>
    <td>[% current_gene.gene.display_name %]</td>
    <td>[% current_gene.gene.location.start %]</td>
    <td>[% current_gene.gene.location.end %]</td>
    <td>[% current_gene.gene.location.strand %]</td>
</tr>
[% END %]
</table>

[% END %]

[% problem_description = problem_data.problem_genes %]

[% IF problem_description %]

<h1>Problem genes</h1>

<p>
  There were [% problem_description.size %] genes with problems:
</p>

<table>

<tr>
  <th></th>
  <th>Error message</th>
  <th>Start</th>
  <th>End</th>
  <th>Strand</th>
</tr>

  [%- FOREACH current_problem_description = problem_description -%]

  <tr>
    <td>[% loop.count %]/[% loop.size %]</td>
    <td>[% current_problem_description.msg %]</td>
    <td>[% current_problem_description.gene.exons.start %]</td>
    <td>[% current_problem_description.gene.exons.end %]</td>
    <td>[% current_problem_description.gene.exons.strand %]</td>
    <td>[% current_problem_description.gene.exons.get_tag_values('note') %]</td>
  </tr>
  <tr>
    <td></td>
    <td></td>
    <td colspan="4"><pre>[% wrap_text(current_problem_description.gene.exons.cds.seq) %]</pre></td>
  </tr>
  <tr>
    <td></td>
    <td></td>
    <td colspan="4"><pre>[% wrap_text(current_problem_description.gene.exons.cds.translate.seq) %]</pre></td>
  </tr>
  [% END -%]
  </table>

  [% END %]
[% END %]